pkg/policy/policies.go

```diff
@@ -1,36 +1,55 @@
 package policy
 
 import (
+	"sync"
 	"sync/atomic"
 
+	bpf "github.com/aquasecurity/libbpfgo"
+
+	"github.com/aquasecurity/tracee/common"
 	"github.com/aquasecurity/tracee/pkg/events"
 	"github.com/aquasecurity/tracee/pkg/filters"
+	"github.com/aquasecurity/tracee/pkg/logger"
 	"github.com/aquasecurity/tracee/pkg/utils"
 )
 
+const (
+	MaxPolicies   int = 64
+	AllPoliciesOn     = ^uint64(0)
+)
+
 var AlwaysSubmit = events.EventState{
 	Submit: AllPoliciesOn,
 }
 
-// TODO: add locking mechanism as policies will change at runtime
+// TODO: refactor filterEnabledPoliciesMap and filterUserlandPoliciesMap
+// maps to use int (Policy id) as key instead of *Policy.
+// TODO: create a new map with policy name as key to speed up LookupByName()
 type Policies struct {
-	policiesArray             [MaxPolicies]*Policy // underlying filter policies array
-	filterEnabledPoliciesMap  map[*Policy]int      // stores only enabled policies
-	filterUserlandPoliciesMap map[*Policy]int      // stores a reduced map with only userland filterable policies
-
-	uidFilterMin            uint64
-	uidFilterMax            uint64
-	pidFilterMin            uint64
-	pidFilterMax            uint64
-	uidFilterableInUserland bool
-	pidFilterableInUserland bool
-
-	filterableInUserland    uint64 // bitmap of policies that must be filtered in userland
-	containerFiltersEnabled uint64 // bitmap of policies that have at least one container filter type enabled
+	rwmu sync.RWMutex
```

Comment:
Ok, either you do a global lock or a fine grained lock. You want to protect a deep copy of the policies array, which is fine, but you leave other parts unprotected (Count() for example is not thread safe). I would make this entire type thread-safe or thread-unsafe. Currently its half-half.

Reply:
Yep. When I put that I was concerned only with the cloning as Policies object is immutable, considering the upcoming versioning, but you pinpointed correctly: we have public methods reading/writing to inner values. Gonna make it full thread-safe.